code_producers/src/llvm_elements/array_switch.rs

```diff
@@ -0,0 +1,133 @@
+    use crate::llvm_elements::instructions::{
+        create_br, create_call, create_conditional_branch, create_eq, create_return_void,
+        create_store, create_gep, create_load, create_return, create_switch,
+    };
+    use crate::llvm_elements::stdlib::{
+        ASSERT_FN_NAME, CONSTRAINT_VALUE_FN_NAME, CONSTRAINT_VALUES_FN_NAME,
+    };
+    use crate::llvm_elements::LLVMIRProducer;
+    use crate::llvm_elements::types::{bigint_type, bool_type, i32_type, void_type};
+
+
+    pub fn get_load_symbol(index_range: &Range<usize>) -> String {
+        format!("__array_load__{}_to_{}", index_range.start, index_range.end)
+    }
+
+    pub fn get_store_symbol(index_range: &Range<usize>) -> String {
+        format!("__array_store__{}_to_{}", index_range.start, index_range.end)
+    }
+
+    pub fn create_array_load_fn<'a>(producer: &dyn LLVMIRProducer<'a>, index_range: &Range<usize>) {
+        // args: array, index
+        // return: bigint loaded from array
+        let bool_ty = bool_type(producer);
+        let bigint_ty = bigint_type(producer);
+        let i32_ty = i32_type(producer);
+        let args = &[
+            bigint_ty.ptr_type(Default::default()).into(),
+            i32_ty.into(),
+        ];
+
+        let func = create_function(producer, &None, 0, "", &get_load_symbol(index_range), bigint_ty.fn_type(args, false));
+        let main = create_bb(producer, func, "main");
+
+        let arr = func.get_nth_param(0).unwrap().into_pointer_value();
+        let arr_idx = func.get_nth_param(1).unwrap().into_int_value();
+
+        // build the switch cases
+        let mut cases = vec![];
+        for idx in index_range.clone().into_iter() {
+            let case_val = i32_ty.const_int(idx.try_into().unwrap(), false);
+            let case_bb = create_bb(producer, func, format!("case_{}", idx).as_str());
+            producer.set_current_bb(case_bb);
+
+            let ptr = create_gep(producer, arr, &[case_val]).into_pointer_value();
+            let val = create_load(producer, ptr).into_int_value();
+            create_return(producer, val);
+
+            cases.push((case_val, case_bb));
+        }
+        // -- else case
+        let else_bb = create_bb(producer, func, "else");
+        producer.set_current_bb(else_bb);
+        create_call(producer, ASSERT_FN_NAME, &[bool_ty.const_zero().into()]);
+        create_return(producer, bigint_ty.const_zero());
+
+        producer.set_current_bb(main);
+
+        create_switch(producer, arr_idx, else_bb, &cases);
+
+    }
+
+    pub fn create_array_store_fn<'a>(producer: &dyn LLVMIRProducer<'a>, index_range: &Range<usize>) {
+        // args: array, index, value
+        // return: void
+        let bool_ty = bool_type(producer);
+        let bigint_ty = bigint_type(producer);
+        let i32_ty = i32_type(producer);
+        let void_ty = void_type(producer);
+        let args = &[
+            bigint_ty.ptr_type(Default::default()).into(),
```

Comment:
nit: This I think should be a array of size 0 instead of a pointer. Same in `create_array_load_fn`

Reply:
This actually took a little re-jiggering (see latest), let me know if I understood what you meant properly